OpenRA.Game/Support/ConditionExpression.cs

```diff
@@ -25,47 +25,105 @@ public class ConditionExpression
+
+			public TokenType(int precedence, Associativity associativity = Associativity.Left)
+			{
+				Precedence = precedence;
+				Associativity = associativity;
+			}
+		}
+
 		class Token
 		{
 			public readonly string Symbol;
 			public readonly int Index;
-			public readonly int Precedence;
-			public readonly Associativity Associativity;
+			public readonly TokenType Type;
+			public int Precedence { get { return Type.Precedence; } }
+			public Associativity Associativity { get { return Type.Associativity; } }
 
-			public Token(string symbol, int index, Associativity associativity, int precedence)
+			public Token(string symbol, int index, TokenType type)
 			{
 				Symbol = symbol;
 				Index = index;
-				Associativity = associativity;
-				Precedence = precedence;
+				Type = type;
 			}
 		}
 
+		class BinaryOperationTokenType : TokenType
+		{
+			public BinaryOperationTokenType(int precedence = 0, Associativity associativity = Associativity.Left)
+				: base(precedence, associativity) { }
+		}
+
 		class BinaryOperationToken : Token
 		{
-			public BinaryOperationToken(string symbol, int index, Associativity associativity = Associativity.Left, int precedence = 0)
-				: base(symbol, index, associativity, precedence) { }
+			public BinaryOperationToken(string symbol, int index, BinaryOperationTokenType type)
+				: base(symbol, index, type) { }
+		}
+
+		class UnaryOperationTokenType : TokenType
+		{
+			public UnaryOperationTokenType(int precedence = 1, Associativity associativity = Associativity.Right)
+				: base(precedence, associativity) { }
 		}
 
 		class UnaryOperationToken : Token
 		{
-			public UnaryOperationToken(string symbol, int index, Associativity associativity = Associativity.Right, int precedence = 1)
-				: base(symbol, index, associativity, precedence) { }
+			public UnaryOperationToken(string symbol, int index, UnaryOperationTokenType type)
+				: base(symbol, index, type) { }
+		}
+
+		class OpenParenToken : Token
+		{
+			public static new readonly TokenType Type = new TokenType(-1);
+			public OpenParenToken(int index) : base("(", index, Type) { }
+		}
+
+		class CloseParenToken : Token
+		{
+			public static new readonly TokenType Type = new TokenType(-1);
+			public CloseParenToken(int index) : base(")", index, Type) { }
 		}
 
-		class OpenParenToken : Token { public OpenParenToken(int index) : base("(", index, Associativity.Left, -1) { } }
-		class CloseParenToken : Token { public CloseParenToken(int index) : base(")", index, Associativity.Left, -1) { } }
 		class VariableToken : Token
 		{
-			public VariableToken(int index, string symbol)
-				: base(symbol, index, Associativity.Left, 0) { }
+			public static new readonly TokenType Type = new TokenType(0);
+			public VariableToken(int index, string symbol) : base(symbol, index, Type) { }
+		}
+
+		class AndToken : BinaryOperationToken
+		{
+			public static new readonly BinaryOperationTokenType Type = new BinaryOperationTokenType();
```

Comment:
Why are these duplicated everywhere instead of being defined on `BinaryOperationToken` etc?